### PR TITLE
Update the integration of libCEED to work with current main (libCEEDv0.10).

### PR DESCRIPTION
**Pull request by YohannDudouit**
Due to conflicting function names, and recent interface changes, the `main` branch of libCEED does not compile with MFEM. This PR adds a version requirement for libCEED v0.9.1 (also working with the current state of the `main` branch)

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2569](https://github.com/mfem/mfem/pull/2569) | @YohannDudouit | @tzanio | @tzanio + @v-dobrev | 3/14/22 | 3/15/22 | 3/17/22 | |

<details>
<summary>PR Checklist</summary>
    
- [x] Code builds.
- [x] Code passes `make style`.
- [x] Update `CHANGELOG`
- [x] Update `INSTALL`:
    - [x] Have the version ranges for any required or optional libraries changed?
- [x] Run `make unittest` to make sure all unit tests pass.
- [x] Run the tests in `tests/scripts`.

</details>

### CODE REVIEW

**Comment by jedbrown** [comment]
Does this need a review or is the issue that libceed needs to release first for this to be merged?

**Comment by v-dobrev** [comment]
If this is required for the libCEED integration in MFEM to work with the libCEED version that will be part of CEED 5.0, then we should merge this before the MFEM v4.4 release -- that's the version we plan to use for CEED 5.0.

**Comment by jedbrown** [comment]
We can help wrap this up on Monday with another needed update.

**Comment by YohannDudouit** [comment]
@jedbrown We updated our libCEED integration with libCEED `main`, do you know which minimal version we should ask in `INSTALL.md`? 0.10?

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by jeremylt** [comment]
We'll be releasing v0.10

File: fem/ceed/util.hpp [hunk 1]
```diff
@@ -17,6 +17,9 @@
 #include <ceed.h>
 #include <ceed/hash.h>
 #include <ceed/backend.h>  // for CeedOperatorField
+#if !CEED_VERSION_GE(0,10,0)
+#error MFEM requires a libCEED version superior to 0.10.0
```
Comment (v-dobrev) [inline fem/ceed/util.hpp:21]: I'm not sure "superior to" is the right term to use here -- use either "greater than or equal to" or simply ">=".

Comment (YohannDudouit) [reply fem/ceed/util.hpp:21]:
```suggestion
#error MFEM requires a libCEED version >= 0.10.0
```

**Review by tzanio — APPROVED** [review]

**Comment by tzanio** [comment]
@jedbrown and @jeremylt -- does this look good from your perspective?

**Review by v-dobrev — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Review by jedbrown — APPROVED** [review]
These changes look reasonable to me. Can MFEM use 64-bit indices? I see some raw `int`s.

Do you want to support matrix assembly via libceed? If you're handing off to PETSc, the arrays can be passed directly to `MatSetPreallocationCOO` and `MatSetValuesCOO`. (We think this is a good standard for on-device matrix assembly and hopefully other libs will do it too.)

https://gitlab.com/micromorph/ratel/-/blob/main/src/ratel-petsc-ops.c#L104-107
https://gitlab.com/micromorph/ratel/-/blob/main/src/ratel-petsc-ops.c#L133-135

**Comment by YohannDudouit** [comment]
> These changes look reasonable to me. Can MFEM use 64-bit indices? I see some raw `int`s.
> 
> Do you want to support matrix assembly via libceed? If you're handing off to PETSc, the arrays can be passed directly to `MatSetPreallocationCOO` and `MatSetValuesCOO`. (We think this is a good standard for on-device matrix assembly and hopefully other libs will do it too.)
> 
> https://gitlab.com/micromorph/ratel/-/blob/main/src/ratel-petsc-ops.c#L104-107 https://gitlab.com/micromorph/ratel/-/blob/main/src/ratel-petsc-ops.c#L133-135

I think this would be interesting to add, I opened an [issue](https://github.com/mfem/mfem/issues/2897) and try to integrate this into MFEM when possible.

**Comment by v-dobrev** [comment]
> Can MFEM use 64-bit indices? I see some raw `int`s.

MFEM uses HYPRE's big-int type for big (global) offsets, indices, sizes, etc. So with suitable HYPRE configuration we can use 64-bit indices in the most important places, at least. There were some discussions to add support for 64-bit ints in other places too (like the `I` and/or `J` arrays in the CSR format) but we don't have that yet.

**Comment by jedbrown** [comment]
Cool. I've wanted to revisit this for PETSc, to promote the `I` indices of CSR by default because it's getting easy (especially when running on big-memory GPUs) to create sequential matrices with more than 2B entries (also for factors created within a sparse direct solver), despite the vector lengths being much smaller. https://gitlab.com/petsc/petsc/-/merge_requests/3186

**Comment by v-dobrev** [comment]
Yes, I've also run into the issue with direct solver (UMFPACK) failing with 32-bit indices on not-so-large systems. That's why in MFEM's wrapper for UMFPACK we have a runtime flag to switch to the `long` version of the solver instead of the `int` version: https://github.com/mfem/mfem/blob/a8b5004ee9c81ae6ee730ae1fc8a3a78625022b1/linalg/solvers.hpp#L1039-L1041

**Comment by jedbrown** [comment]
Ah, good thought. This solver natively uses pointer-sized offsets and meets/beats CHOLMOD across a suite of problems. https://gitlab.com/hodge_star/catamari

**Pull request merged**